README.md
<p align="center">
  <a href="https://adshares.net/">
    <img src="https://adshares.net/logos/ads.svg" alt="Adshares" width=72 height=72>
  </a>
  <h3 align="center"><small>Adshares DCL plugin</small></h3>
  <p align="center">
    <a href="https://github.com/adshares/decentraland/issues/new?template=bug_report.md&labels=Bug">Report bug</a>
    ·
    <a href="https://github.com/adshares/decentraland/issues/new?template=feature_request.md&labels=New%20Feature">Request feature</a>
    ·
    <a href="https://docs.adshares.net/adserver/">Docs</a>
  </p>
</p>

<br>

Adshares plugin for **Decentraland** SDK. You can use it in your scenes for addition hassle-free earnings.

Easy configuration. Just fill out your payout address (Adshares blockchain or as BEP20 token) and get paid in $ADS for every visitor on your scene. See [DCL Example Scene](https://github.com/adshares/dcl-scene).

## Install

```bash
npm install @adshares/decentraland -B
```

### Update

```bash
npm update @adshares/decentraland
```

## Usage

### 1. Import dependencies

```js
import * as Ads from '@adshares/decentraland'
```

### 2. Create supply agent

```js
const agent = new Ads.SupplyAgent(adserver: string, publisherId: string)
```

or

```js
const agent = Ads.SupplyAgent.fromWallet(adserver: string, chain: string, wallet: string)
```

In function **fromWallet()** first argument is adserver network, second argument is payout network (ads, bsc), and third argument is wallet address in this payout network.

### 3. Create Placement or Stand

#### Placements

Placements are the basic objects for displaying creatives.
Every placement implements IEntity and has access to IEntity methods.

Usage:

```js
const placement = new Ads.PlainPlacement(name: string, options?: {})
```

Available options:

```js
{
  position?: Vector3, // @decentraland-ecs
  rotation?: Quaternion, // @decentraland-ecs
  width?: number,
  ratio?: '9:16' | '3:4' | '1:1' | '4:3' | '16:9',
  no?: number | null,
  types?: string[] | null,
  mimes?: string[] | null,
}
```

#### UI placements

UI placements are elements of interface for displaying creatives, when player is on scene.
Keep in might that UI placement doesn't need to be added to Decentraland engine. UI placement can be closed after 5 seconds from displaying.

Usage:

```js
const placement = new Ads.UIPlacement(name: string, position: 'top' | 'bottom' | 'left' | 'right' | 'center')
```

#### Stands

Stands combine several placements into one object.
Every stand implements IEntity and has access to IEntity methods.

The simplest stand is the PlainStand, which repeats the PlainPlacement and adds a frame and a background that covers the gaps between the layers. You can define own Material for frame.

```js
const plain = new Ads.PlainStand(name: string, options?: {})
```
Available options:

```js
{
  position?: Vector3, // @decentraland-ecs
  rotation?: Quaternion, // @decentraland-ecs
  width?: number,
  ratio?: '9:16' | '3:4' | '1:1' | '4:3' | '16:9',
  no?: number | null,
  types?: string[] | null,
  mimes?: string[] | null,
  frameMaterial?: Material,
}
```

Available stands:

| Class     | Description                                                      | Ratios                                    | Models                                                                                                                              |
|-----------|------------------------------------------------------------------|-------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------|
| Totem     | A simple advertising totem                                       | 9:16                                      | ads_totem_9_16.glb                                                                                                                  |
| Citylight | Rectangular 4-sided advertising totem with the ability to rotate | 9:16                                      | ads_citylight_9_16.glb                                                                                                              |
| Billboard | Advertising billboard with multiple ratios                       | 9:16<br/> 16:9<br/> 3:4<br/> 4:3<br/> 1:1 | ads_billboard_9_16.glb<br/> ads_billboard_16_9.glb<br/> ads_billboard_3_4.glb<br/> ads_billboard_4_3.glb<br/> ads_billboard_1_1.glb | 


You must copy the models to the root scene folder before using the stands. It's a good practice to copy only the models you use.

```bash
mkdir -p models/@adshares && cp -r node_modules/@adshares/decentraland/models/ads_totem_9_16.glb models/@adshares
```

Usage Totem:

```js
const totem = new Ads.Totem(name: string, options?: {})
```

Available options:

```js
{
  types?: string[] | null,
  mimes?: string[] | null,
}
```

Usage Citylight:

```js
const citylight = new Ads.Citylight(name: string, options?: {})
```

Available options:

```js
{
  types?: string[] | null,
  mimes?: string[] | null,
  rotation?: boolean, // default true
}
```

Usage Billboard:

```js
const billboard = new Ads.Billboard(name: string, options?: {})
```

Available options:

```js
{
  types?: string[] | null,
  mimes?: string[] | null,
  ratio?: '9:16' | '3:4' | '1:1' | '4:3' | '16:9', // default 1:1
}
```


### 4. Add placements into Decentraland engine

```js
engine.addEntity(placement)
// or
engine.addEntity(totem)
```
or 

```js
placement.setParent(myEntity)
// or
totem.setParent(myEntity)
```

### 5. Add placement into agent and spawn banner

```js
agent.addPlacement(placement: IPlacement | IStand).spawn()
```

### Example to add placements

```js
import * as Ads from '@adshares/decentraland'

const agent = Ads.SupplyAgent.fromWallet('https://app.web3ads.net', 'ads', '0001-00000000-9B6F')

const placement1 = new Ads.PlainPlacement('unit1', {
  position: new Vector3(8, 2, 8),
  rotation: new Quaternion(0, 0, 0, 1),
  width: 5,
  ratio: '16:9',
})
engine.addEntity(placement1)

const placement2 = new Ads.PlainPlacement('unit2', {
  position: new Vector3(11, 2, 6),
  rotation: new Quaternion(0, 1, 0, 1),
  width: 3,
  ratio: '3:4',
})
engine.addEntity(placement2)

agent.addPlacement(placement1, placement2, totem).spawn()
```

![Placement example](/assets/placement_example.png "Decentraland scene")


### Example to add advertising stands

```js
import * as Ads from '@adshares/decentraland'

const agent = Ads.SupplyAgent.fromWallet('https://app.web3ads.net', 'ads', '0001-00000000-9B6F')

const totem = new Ads.Totem('totem')
totem.addComponent(new Transform({
  position: new Vector3(3,0,8),
}))
engine.addEntity(totem)

const billboard = new Ads.Billboard('billboard_16_9', { ratio: '16:9' })
billboard.addComponent(new Transform({
  position: new Vector3(11,0,8),
}))
engine.addEntity(billboard)
agent.addPlacement(totem, billboard).spawn()
```

![Stand example](/assets/stands_example.png "Decentraland scene") 

### Example to add UI placement

```js
import * as Ads from '@adshares/decentraland'

const agent = Ads.SupplyAgent.fromWallet('https://app.web3ads.net', 'ads', '0001-00000000-9B6F')

const uiPlacementRight = new UIPlacement('UI-right', 'right')
const uiPlacementLeft = new UIPlacement('UI-left', 'left')
const uiPlacementCenter = new UIPlacement('UI-center', 'center')

agent.addPlacement(uiPlacementRight, uiPlacementLeft, uiPlacementCenter).spawn()
```

![UI placement example](/assets/ui_placement_example.png "Decentraland scene")

### Contributing

Please follow our [Contributing Guidelines](docs/CONTRIBUTING.md)

### Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/adshares/decentraland/tags).

## Authors

- **[Mykola Zhura](https://github.com/Niko-Yea)** - _programmer_
- **[Maciej Pilarczyk](https://github.com/m-pilarczyk)** - _programmer_
- **[Contributor](https://github.com/smartsir796)** - _programmer_

See also the list of [contributors](https://github.com/adshares/decentraland/contributors) who participated in this project.

### License

This project is licensed under the GNU General Public License v3.0 - see the [LICENSE](LICENSE) file for details.

## More info

- [DCL Example Scene](https://github.com/adshares/dcl-scene)
- [DCL Smart Banner](https://github.com/adshares/dcl-smart-banner)
- [Adshares](https://adshares.net)
- [Adshares Docs](https://docs.adshares.net)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.3.1] - 2022-07-06
### Fixed
- npm build

## [2.3.0] - 2022-05-30
### Added
- Base placement without any side effect
- Plain stand with frame and include base placement
### Fixed
- Blinking banner
- Display placement when banner not found

## [2.2.0] - 2022-04-26
### Added
- UI placements support
### Fixed
- Creative is too close to background on plain placement

## [2.1.1] - 2022-04-25
### Fixed
- Fix citylight rotation system, when entity has no transform component
- Fix models & src folders

## [2.1.0] - 2022-03-27
### Added
- Stands support
- Totem, Citylight and Billboard stands
### Changed
- Build script

## [2.0.2] - 2022-03-07
### Fixed
- Remove entity from engine if exist

## [2.0.1] - 2022-03-03
### Fixed
- Adshares logo
- Error handling

## [2.0.0] - 2022-03-03
### Changed
- Code refactor
- Support for new AdServer API
### Removed
- Smart item code (moved to the separate project)

## [1.1.8] - 2022-11-08
### Fixed
- Less invasive error rendering

## [1.1.7] - 2022-06-20
### Changed
- New SDK version
- Builder compatibility
- Normalize ad server URL

## [1.1.6] - 2022-06-20
### Changed
- Fix scaling for nested ads
- Allow to customize screen material
- Video ad sound plays only after click
- Refresh ads periodically
- Support for `info_box` flag

## [1.1.5] - 2022-05-23
### Changed
- Remove dependency on dcl builder scripts

## [1.1.4] - 2022-04-25
### Changed
- Use signedFetch for logging events

## [1.1.3] - 2022-04-04
### Changed
- Npm package setup

## [1.1.2] - 2022-03-22
### Changed
- Add account and metamask to context

## [1.1.1] - 2022-03-15
### Fixed
- Optimize ads loading

## [1.1.0] - 2022-03-01
### Added
- Video ads support

[Unreleased]: https://github.com/adshares/decentraland/compare/v2.3.1...HEAD
[2.3.1]: https://github.com/adshares/decentraland/compare/v2.3.0...2.3.1
[2.3.0]: https://github.com/adshares/decentraland/compare/v2.2.0...2.3.0
[2.2.0]: https://github.com/adshares/decentraland/compare/v2.1.1...2.2.0
[2.1.1]: https://github.com/adshares/decentraland/compare/v2.1.0...2.1.1
[2.1.0]: https://github.com/adshares/decentraland/compare/v2.0.2...2.1.0
[2.0.2]: https://github.com/adshares/decentraland/compare/v2.0.1...2.0.2
[2.0.1]: https://github.com/adshares/decentraland/compare/v2.0.0...2.0.1
[2.0.0]: https://github.com/adshares/decentraland/compare/v1.1.8...2.0.0
[1.1.8]: https://github.com/adshares/decentraland/compare/v1.1.7...1.1.8
[1.1.7]: https://github.com/adshares/decentraland/compare/v1.1.6...1.1.7
[1.1.6]: https://github.com/adshares/decentraland/compare/v1.1.5...1.1.6
[1.1.5]: https://github.com/adshares/decentraland/compare/v1.1.4...v1.1.5
[1.1.4]: https://github.com/adshares/decentraland/compare/v1.1.3...v1.1.4
[1.1.3]: https://github.com/adshares/decentraland/compare/v1.1.2...v1.1.3
[1.1.2]: https://github.com/adshares/decentraland/compare/v1.1.1...v1.1.2
[1.1.1]: https://github.com/adshares/decentraland/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/adshares/decentraland/releases/tag/v1.1.0


src/item.ts
import {getUserAccount} from '@decentraland/EthereumController'
import {getParcel, ILand} from '@decentraland/ParcelIdentity'
import {setTimeout} from './timer'

let SignedFetch: Function
let isBuilder = false

async function importFetch(): Promise<any> {
    if (SignedFetch) {
        return SignedFetch
    }

    await import("@decentraland/SignedFetch").then((x: any) => {
        if (typeof x === 'object') {
            x = x[0]
        }
        SignedFetch = x.signedFetch
    }, x => {
        SignedFetch = fetch
        isBuilder = true
    })
    return SignedFetch
}

export type Props = {
    payout_network: string
    payout_address: string
    keywords: string
    zone_name: string
    adserver: string
    exclude: string
    hide_errors?: boolean
    onMaterial?: (material: Material) => void // use to customize ad sceeen material properties
}

const b64ch =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_='
const b64chs = Array.prototype.slice.call(b64ch)

const UrlSafeBase64Encode = (bin: string) => {
    bin = unescape(encodeURIComponent(bin))
    let u32, c0, c1, c2, asc = ''
    const pad = bin.length % 3
    for (let i = 0; i < bin.length;) {
        if ((c0 = bin.charCodeAt(i++)) > 255 ||
            (c1 = bin.charCodeAt(i++)) > 255 ||
            (c2 = bin.charCodeAt(i++)) > 255) {
            throw new TypeError('invalid character found')
        }
        u32 = (c0 << 16) | (c1 << 8) | c2
        asc += b64chs[u32 >> 18 & 63]
            + b64chs[u32 >> 12 & 63]
            + b64chs[u32 >> 6 & 63]
            + b64chs[u32 & 63]
    }
    return pad ? asc.slice(0, pad - 3) : asc
}

const addUrlParam = function (url: string, names: any, value: any = null) {
    if (typeof names != 'object') {
        const tmp = names
        names = {}
        names[tmp] = value
    }
    for (let name in names) {
        value = names[name]
        let param = name + '=' + encodeURIComponent(value)
        const qPos = url.indexOf('?')
        if (qPos > -1) {
            url += (qPos < url.length ? '&' : '') + param
        } else {
            url += '?' + param
        }
    }
    return url
}

interface IHash {
    [details: string]: boolean;
}

export default class AdsharesBanner {
    impressionId: string = ''
    bannerCounter: number = 0
    loadedAdusers: IHash = {}

    init(args?: any) {
    }

    getRandId(bytes: Number): string {
        let d = new Date().getTime()

        let chars = []
        for (let i = 0; i < bytes; i++) {
            const r = (d + Math.random() * 256) % 256 | 0
            d = Math.floor(d / 256)
            chars.push(String.fromCharCode(r))
        }

        return chars.join('')
    }

    getImpressionId(): string {
        if (this.impressionId == '') {
            this.impressionId = UrlSafeBase64Encode(this.getRandId(16))
        }
        return this.impressionId
    }

    getSceneId(land: ILand): string {
        return 'scene-' +
            land.sceneJsonData.scene.base.replace(new RegExp('-', 'g'), 'n')
                .replace(',', '-')
    }

    getSceneTags(land: ILand, extraTags: string[]): string {
        if (land.sceneJsonData.tags) {
            extraTags = extraTags.concat(land.sceneJsonData.tags)
        }
        return extraTags.join(',')
    }

    getCid() {
        let i, l, n
        let s = this.getRandId(15) + '\0'
        let o = ''
        for (i = 0, l = s.length; i < l; i++) {
            n = s.charCodeAt(i).toString(16)
            o += n.length < 2 ? '0' + n : n
        }
        return o
    }

    getCombinedScale(host: Entity): Vector3 {
        let scale = host.getComponent(Transform).scale
        let entity = host.getParent();

        while (entity) {
            if (entity.hasComponent(Transform)) {
                let pScale = entity.getComponent(Transform).scale

                scale = scale.multiply(pScale);

            }
            entity = entity.getParent();
        }

        return scale
    }

    async find(host: Entity, props: Props, cleanup: boolean = false) {
        const userAccount = await getUserAccount()
        const parcel = await getParcel()
        const scale = this.getCombinedScale(host)

        let signedFetch = await importFetch()

        if (this.impressionId == '') {
            const register_url = props.adserver + '/supply/register?iid=' + this.getImpressionId()
            signedFetch(register_url + '&stid=' + userAccount).then()
            this.loadedAdusers[register_url] = true
        }

        let request = {
            pay_to: props.payout_network + ':' + props.payout_address,
            view_id: this.getImpressionId(),
            zone_name: props.zone_name,
            width: scale.x,
            height: scale.y,
            min_dpi: 10,
            exclude: JSON.parse(props.exclude),
            type: ['image', 'video'],
            mime_type: ['image/jpeg', 'image/png', 'video/mp4'],
            context: {
                site: {
                    url: 'https://' + this.getSceneId(parcel.land) + '.decentraland.org/',
                    keywords: this.getSceneTags(parcel.land, props.keywords.split(',')),
                    metamask: 1,
                },
                user: {
                    account: userAccount,
                },
            },
            medium: 'metaverse',
            vendor: 'decentraland',
            version: '1.1.7',
        }

        let response: any = {}

        try {
            let callUrl = props.adserver + '/supply/anon?stid=' + userAccount
            let callResponse = await fetch(callUrl, {
                headers: {'Content-Type': 'application/json'},
                method: 'POST',
                body: JSON.stringify(request),
            })
            response = await callResponse.json()
            log(request, response)
        } catch (e) {
            log('failed to reach URL', e)
        }
        if(cleanup) {
            this.clearChildren(host)
        }
        let banner
        if (response.banners) {
            banner = response.banners[0]
            if (banner) {
                banner.cid = this.getCid()
                let viewContext = {
                    page: {
                        iid: request.view_id,
                        url: request.context.site.url,
                        keywords: request.context.site.keywords,
                        metamask: request.context.site.metamask,
                    },
                    user: {
                        account: request.context.user.account,
                    },
                }
                banner.click_url = addUrlParam(banner.click_url,
                    {
                        cid: banner.cid,
                        ctx: UrlSafeBase64Encode(JSON.stringify(viewContext)),
                        iid: request.view_id,
                        stid: userAccount,
                    },
                )
                banner.view_url = addUrlParam(banner.view_url,
                    {
                        cid: banner.cid,
                        ctx: UrlSafeBase64Encode(JSON.stringify(viewContext)),
                        iid: request.view_id,
                        json: 1,
                        stid: userAccount,
                    })
                await this.renderBanner(host, props, banner)

                if (false !== banner.info_box) {
                    this.showWaterMark(host, props, request, banner)
                }

                try {
                    let loadedAdusers = this.loadedAdusers
                    signedFetch(banner.view_url).then(function (response: any) {
                        let object;
                        if (response.text) {
                            object = JSON.parse(response.text);
                        } else {
                            object = response.json;
                        }
                        if (object.aduser_url && !loadedAdusers[object.aduser_url]) {
                            signedFetch(object.aduser_url)
                            loadedAdusers[object.aduser_url] = true
                        }
                    })

                } catch (e) {
                    log('view log failed', e)
                }
            } else {
                this.renderText(host, props, 'https://assets.adshares.net/metaverse/notfound.png',
                    'Banner not found\n\nImpression ID: ' + request.view_id +
                    '\n\nconfig: ' + JSON.stringify(props, null, '\t'))
            }
        }
        if (!response.success) {
            let errors: string[] = []
            if (response.errors) {
                let k: any
                let v: any
                for (k in response.errors) {
                    //errors.push(k)
                    v = response.errors[k]
                    if (typeof v !== 'object') {
                        v = Array(v)
                    }
                    v.forEach((text: string) => {
                        errors.push(text)
                    })
                    errors.push('')
                }

            }
            errors.push('\nconfig: ' + JSON.stringify(props, null, '\t'))
            this.renderError(host, props, errors)
        }


        setTimeout(() => {
            this.find(host, props, !isBuilder)
        }, banner && banner.refresh ? banner.refresh : 30000)
    }

    clearChildren(host: Entity) {
        for (let k in host.children) {
            let entity = host.children[k]
            entity.removeComponent(Transform)
            entity.removeComponent(PlaneShape)
            entity.removeComponent(Material)
            entity.removeComponent(OnClick)
            engine.removeEntity(host.children[k])
            delete host.children[k]
        }
    }

    showWaterMark(host: Entity, props: Props, request: any, banner: any) {
        const hostScale = this.getCombinedScale(host)
        let url = 'https://assets.adshares.net/metaverse/watermark.png'

        let QRTexture = new Texture(url)

        let size = Math.sqrt(hostScale.x * hostScale.y) / 10
        let scale = {
            x: size / hostScale.x,
            y: size / hostScale.y,
        }

        let QRPlane = new Entity()
        QRPlane.setParent(host)
        QRPlane.addComponent(new PlaneShape())
        QRPlane.addComponent(
            new Transform({
                position: new Vector3(-0.5 + scale.x / 2, 1 - scale.y / 2, 0.01),
                rotation: Quaternion.Euler(180, 0, 0),
                scale: new Vector3(scale.x, scale.y, 1),
            }),
        )

        let QRMaterial = new Material()

        QRMaterial.metallic = 0
        QRMaterial.roughness = 1
        QRMaterial.specularIntensity = 0
        QRMaterial.albedoTexture = QRTexture

        if (props.onMaterial) {
            props.onMaterial(QRMaterial)
        }

        QRPlane.addComponent(QRMaterial)

        QRPlane.addComponent(
            new OnClick(() => {
                let url = addUrlParam(props.adserver + '/supply/why', {
                    bid: banner.id,
                    cid: banner.cid,
                    iid: request.view_id,
                    url: banner.serve_url,
                    ctx: UrlSafeBase64Encode(JSON.stringify(request.context)),
                })
                openExternalURL(url)
            }),
        )
    }

    renderText(host: Entity, props: Props, icon: string, message: string) {
        let QRPlane2 = new Entity()
        QRPlane2.setParent(host)
        QRPlane2.addComponent(new PlaneShape())
        QRPlane2.addComponent(
            new Transform({
                position: new Vector3(0, 0.5, 0),
                rotation: Quaternion.Euler(180, 0, 0),
                scale: new Vector3(1, 1, 1),
            }),
        )

        let QRMaterial2 = new Material()

        QRMaterial2.metallic = 0
        QRMaterial2.roughness = 1
        QRMaterial2.specularIntensity = 0

        QRMaterial2.albedoColor = Color3.White()
        QRPlane2.addComponent(QRMaterial2)

        let QRPlane = new Entity()
        QRPlane.setParent(host)
        QRPlane.addComponent(new PlaneShape())

        const hostScale = this.getCombinedScale(host)
        let size = Math.sqrt(hostScale.x * hostScale.y) / 2
        let scale = {
            x: size / hostScale.x,
            y: size / hostScale.y,
        }

        QRPlane.addComponent(
            new Transform({
                position: new Vector3(0, 0.5, 0.01),
                rotation: Quaternion.Euler(180, 180, 0),
                scale: new Vector3(scale.x, scale.y, 1),
            }),
        )

        let QRMaterial = new Material()
        QRMaterial.albedoColor = Color3.White()
        QRMaterial.metallic = 0
        QRMaterial.roughness = 1
        QRMaterial.specularIntensity = 0

        let QRTexture = new Texture(icon)
        QRMaterial.albedoTexture = QRTexture
        QRPlane.addComponent(QRMaterial)

        QRPlane.addComponent(
            new OnPointerDown(() => {
                const canvas = new UICanvas()
                canvas.visible = false

                const textInput = new UIInputText(canvas)
                textInput.width = '30%'
                textInput.height = '70%'
                textInput.vAlign = 'center'
                textInput.hAlign = 'right'
                textInput.fontSize = 10
                textInput.paddingLeft = textInput.paddingRight = textInput.paddingTop = textInput.paddingBottom = 10
                textInput.placeholder = message
                textInput.color = Color4.White()
                textInput.positionX = '-5%'
                textInput.isPointerBlocker = false
                textInput.hTextAlign = 'left'
                textInput.vTextAlign = 'top'
                canvas.visible = true
            }),
        )
    }

    renderError(host: Entity, props: Props, errors: string[]) {
        this.renderText(host, props, 'https://assets.adshares.net/metaverse/error.png',
            'Banner ERROR\n\n' + errors.join('\n'))
    }

    async renderBanner(host: Entity, props: Props, banner: any) {
        let QRPlane = new Entity()
        QRPlane.setParent(host)
        QRPlane.addComponent(new PlaneShape())
        QRPlane.addComponent(
            new Transform({
                position: new Vector3(0, 0.5, 0),
                rotation: Quaternion.Euler(banner.type == 'image' ? 180 : 0, 0, 0),
                scale: new Vector3(1, 1, 1),
            }),
        )

        let QRMaterial = new Material()

        QRMaterial.metallic = 0
        QRMaterial.roughness = 1
        QRMaterial.specularIntensity = 0
        let QRTexture: Texture | VideoTexture
        if (banner) {
            if (banner.type == 'image') {
                QRTexture = new Texture(banner.serve_url)
                QRMaterial.albedoTexture = QRTexture
            } else if (banner.type == 'video') {
                let video_url = banner.serve_url
                video_url += video_url.indexOf('?') == -1 ? '?' : '&'
                const video = new VideoClip(video_url + '/y.mp4')
                QRTexture = new VideoTexture(video)
                QRTexture.loop = true
                QRTexture.volume = 0
                QRMaterial.albedoTexture = QRTexture
                QRTexture.play()
            } else {
                this.renderError(host, props, ['Invalid banner format: ' + banner.type])
            }

        } else {
            QRMaterial.albedoColor = Color3.White()
        }

        if (props.onMaterial) {
            props.onMaterial(QRMaterial)
        }

        QRPlane.addComponent(QRMaterial)

        if (banner) {
            QRPlane.addComponent(
                new OnClick(() => {
                    if (QRTexture instanceof VideoTexture) {
                        if (QRTexture.volume == 0) {
                            QRTexture.volume = 1;
                            return
                        } else {
                            QRTexture.volume = 0;
                        }
                    }
                    openExternalURL(banner.click_url)
                }),
            )
        }
    }

    normalizeProps(props: Props) {
        while(props.adserver.slice(-1) == '/') {
            props.adserver = props.adserver.slice(0, -1)
        }
    }

    spawn(host: Entity, props: Props, channel: any = null) {
        this.normalizeProps(props)
        this.bannerCounter++
        if (this.bannerCounter > 20) {
            this.renderError(host, props, ['To many banners, max 20'])
        } else {
            this.find(host, props).then()
        }
    }
}
